contributors/design-proposals/storage/local-storage-pv.md

```diff
@@ -761,3 +761,202 @@ The central manager will poll each worker for their discovered volumes and creat
+
+For LVM, the proposed workflow is:
+
+1. Check if there is already a PV pre-bound to the PVC, if so, return directly
+(This can happen when a volume has already been provisioned for a PVC,
+but has not been bound to the PVC by the controller yet).
+2. Check if Storage Class of the PVC is in record of its storageClass map.
+If not, return error.
+3. Check if the class is available for dynamic provisioning (i.e. has an input of `volumeGroup`).
+If not, return error.
+4. Check if `Free Size` of the VG under the class can satisfy the capacity requested
+in the PVC. If not, return error.
+5. Create `logical volume (LV)` object in the VG.
+6. Create a PV API object that is pre-bound to the PVC, and the path
+of local storage in the PV should be the path of the LV created above.
+
+##### Cleanup after Release
+
+Currently we already have a component `deleter` to cleanup volumes/PVs of static local volume,
+detail cleanup actions are managed via custom scripts/binaries that specified in `blockCleanerCommand`.
+
+Existing script `quick_reset.sh` calls `mkfs -F` and `wipefs` to cleanup a block.
+In addition to that, a hook named `volumeDeleteFunc` will be added into `deleter`,
+and for dynamic privisioned volumes, the deleting process should be:
+1. Call scripts/binaries in `blockCleanerCommand` to clean up data in the volume.
+2. Call `volumeDeleteFunc` to recycle the volume behind the PV (e.g. delete the LV for lvm).
+3. Delete the PV object.
+
+##### Expansion of LocalVolumeSource
+
+Currently, the path in `LocalVolumeSource` must be a directory. Thus, for dynamic
+provisioned volumes, the provisioner need to mount the LV to a host path first,
+to allow kubelet for another mount. It increases the complexity of the plugin.
+
+Therefore, in-tree local storage plugin will be updated to support both directory
+and block as a source of mount. The mount process would be:
+1. Detect if given path in volumeSource is mounted or not, if mounted,
+continue to work in the existing way.
+2. If not, Check if global path `{pluginDir}/mounts/{volumeID}` exists.
+and create the directory if not.
+3. Check if the global path is of mount point, if not, mount given path to
+the global path.
+4. Manage `mkfs` for the global path if it is newly created or mounted.
+5. Mount the global path to the path in pods.
+
+New field `FSType` will be added in `LocalVolumeSource` to support `mkfs` in the mounter.
+
+```
+type LocalVolumeSource struct {
+	// The full path to the volume on the node
+	// For alpha, this path must be a directory
+	// Once block as a source is supported, then this path can point to a block device
+	Path string
+	// Filesystem type to mount.
+	// Must be a filesystem type supported by the host operating system.
+	// Ex. "ext4", "xfs", "ntfs". Implicitly inferred to be "ext4" if unspecified.
+	// +optional
+	FSType string
+}
+```
+
+**Note:** The change will only work on PersistentVolumes whose volumeMode are "Filesystem",

```

Comment:
Can you also add a workflow for how dynamic provisioning of Block volumes will work? (I think it will work with this design as long as PV is created with correct volumeMode)

Reply:
Well, when creating PV objects, the provisioner will simply specify its volumeMode according to that in the PVC; and for PV of Block mode, the in-tree local storage plugin simply behavior in existing way.

So I think nothing new will be introduced for PV of Block mode

Reply:
Maybe I should stress the way to specify the volumeMode of PVs in the workflow of provisioning

Reply:
Yeah, I think nothing new needs to be done.  But it may be good to just explicitly show that here for completeness.